File: lib/htuml2txt.pl
#!/usr/local/bin/perl

# converts html files into ascii by just stripping anything between
#  < and >
# written 4/21/96 by Michael Smith for WebGlimpse
#
# Added code to replace html codes for special chars with the
# characters themselves.  12/19/98 --GB
#
# Also add space in place of space-producing HTML tags
# 12/22/98 --GB


BEGIN{
        $WEBGLIMPSE_LIB = '|WEBGLIMPSE_LIB|';
        unshift(@INC, "$WEBGLIMPSE_LIB");  # Find the rest of our libs
} 
use wgFilter;


$SAVE_LINE_MARK = '_~_';


$carry=0;
$lineno = -1;
$incomplete_tag = 0;
$last_incomplete_tag = 0;

@lines = <STDIN>;

&wgFilter::SkipSection('<SCRIPT\s*[^>]*>','<\/SCRIPT>',\@lines);
&wgFilter::SkipSection('<STYLE\s*[^>]*>','<\/STYLE>',\@lines);

foreach $line (@lines){
	$lineno++;	# we need line number in ORIGINAL file for jump-to-line later

	# put returns back in in case user uploaded wrong
	$line =~ s/\r\n/\n/g;

# if we put in extra returns, not every line will start with <linenum>
# take out for now, consider putting back in with <linenum> inserted
#	$line =~ s/\r/\n/g;

	# now take out the ones we don't want!
	$line =~ s/([^>\.\s])\n/$1 /g;
	$line =~ s/(\s)\n/$1/g;

	if($carry==1){
		# remove all until the first >
		next if($line!~s/[^>]*>//);
		# if we didn't do next, it succeeded -- reset carry
		$carry=0;
	} 

	# save <TITLE, <A, <FRAME, <BASE tags
	$line =~ s/<(\/?)(title|a|frame|base)/#~#$1$2/ig;

	while($line=~s/(<[^\s>][^>]*>)/&addspace($1,$lineno)/ge){};

	if($line=~s/<[^\s>].*$//){
		$carry=1;
	}

	# put saved tags back, and check if the tag is complete
	if ($line =~ /#~#[^>]*$/) {
		$incomplete_tag = 1;
	} else {
		$incomplete_tag = 0;
	}
	$line =~ s/#~#(\/?)(title|a|frame|base)/<$1$2/ig;	

	# put lineno tags back
	$line =~ s/$SAVE_LINE_MARK([0-9]+)$SAVE_LINE_MARK/<$1>/g;

	# If we may have a html-encoded char, check and replace with actual char
	if ($line =~ /\&[^;]{2,6};/) {
		$line = &fixspecial($line);
	}
	if ($line) {
		if (! $last_incomplete_tag) {
			print "<$lineno>";
		}
		print $line;
	}
	if (!$last_incomplete_tag || $line =~ />/) {
		$last_incomplete_tag = $incomplete_tag;
#$last_incomplete_tag && warn "Line $line is incomplete\n";
	}
}


sub addspace () {

	$_ = shift;
	my $lineno = shift;
		
	# Check for tags that should NOT return a space.  Common tags first, then group alphabeticallly
	/(<\/?a[\s>])|(<\/?b>)|(<\/?i>)|(<\/?em>)|(<\/?font)/i && return '';
	/(<\/?strong)|(<\/?sup)|(<\/?sub)|(<\/?samp)|(<\/?strike)|(<\/?style)|(<\/?small)/i && return;
	/(<\/?big)|(<\/?base)|(<\/?cite)|(<\/?code)|(<\/?dfn)|(<\/?kbd)|(<\/?link)|(<\/?meta)/i && return;
	/(<\/?tt>)|(<\/?u>)|(<\/?var)/i && return '';

	# Check for tags that need a return for a record break.
	# insert <linenumber> at the beginning of the next line
	/(<\/?p[\s>])|(<\/?br[\s>])|(<\/?tr[\s>])|(<\/?hr[\s>])|(<\/?li[\s>])/i && return "\n$SAVE_LINE_MARK$lineno$SAVE_LINE_MARK";


	# Otherwise, put in a space
	return ' ';

}


sub fixspecial () {

	$_ = shift;

s/\&#160;/ /g;
s/\&nbsp;/ /g;
s/\&#161;/¡/g;
s/\&iexcl;/¡/g;
s/\&#162;/¢/g;
s/\&cent;/¢/g;
s/\&#163;/£/g;
s/\&pound;/£/g;
s/\&#164;/¤/g;
s/\&curren;/¤/g;
s/\&#165;/¥/g;
s/\&yen;/¥/g;
s/\&#166;/¦/g;
s/\&brvbar;/¦/g;
s/\&#167;/§/g;
s/\&sect;/§/g;
s/\&#168;/¨/g;
s/\&uml;/¨/g;
s/\&#169;/©/g;
s/\&copy;/©/g;
s/\&#170;/ª/g;
s/\&ordf;/ª/g;
s/\&#171;/«/g;
s/\&laquo;/«/g;
s/\&#172;/¬/g;
s/\&not;/¬/g;
s/\&#173;/\\/g;
s/\&shy;/\\/g;
s/\&#174;/®/g;
s/\&reg;/®/g;
s/\&#175;/¯/g;
s/\&macr;/¯/g;
s/\&#176;/°/g;
s/\&deg;/°/g;
s/\&#177;/±/g;
s/\&plusmn;/±/g;
s/\&#178;/²/g;
s/\&sup2;/²/g;
s/\&#179;/³/g;
s/\&sup3;/³/g;
s/\&#180;/´/g;
s/\&acute;/´/g;
s/\&#181;/µ/g;
s/\&micro;/µ/g;
s/\&#182;/¶/g;
s/\&para;/¶/g;
s/\&#183;/·/g;
s/\&middot;/·/g;
s/\&#184;/¸/g;
s/\&cedil;/¸/g;
s/\&#185;/¹/g;
s/\&sup1;/¹/g;
s/\&#186;/º/g;
s/\&ordm;/º/g;
s/\&#187;/»/g;
s/\&raquo;/»/g;
s/\&#188;/¼/g;
s/\&frac14;/¼/g;
s/\&#189;/½/g;
s/\&frac12;/½/g;
s/\&#190;/¾/g;
s/\&frac34;/¾/g;
s/\&#191;/¿/g;
s/\&iquest;/¿/g;
s/\&#192;/À/g;
s/\&Agrave;/À/g;
s/\&#193;/Á/g;
s/\&Aacute;/Á/g;
s/\&#194;/Â/g;
s/\&circ;/Â/g;
s/\&#195;/Ã/g;
s/\&Atilde;/Ã/g;
s/\&#196;/Ä/g;
s/\&Auml;/Ä/g;
s/\&#197;/Å/g;
s/\&ring;/Å/g;
s/\&#198;/Æ/g;
s/\&AElig;/Æ/g;
s/\&#199;/Ç/g;
s/\&Ccedil;/Ç/g;
s/\&#200;/È/g;
s/\&Egrave;/È/g;
s/\&#201;/É/g;
s/\&Eacute;/É/g;
s/\&#202;/Ê/g;
s/\&Ecirc;/Ê/g;
s/\&#203;/Ë/g;
s/\&Euml;/Ë/g;
s/\&#204;/Ì/g;
s/\&Igrave;/Ì/g;
s/\&#205;/Í/g;
s/\&Iacute;/Í/g;
s/\&#206;/Î/g;
s/\&Icirc;/Î/g;
s/\&#207;/Ï/g;
s/\&Iuml;/Ï/g;
s/\&#208;/Ð/g;
s/\&ETH;/Ð/g;
s/\&#209;/Ñ/g;
s/\&Ntilde;/Ñ/g;
s/\&#210;/Ò/g;
s/\&Ograve;/Ò/g;
s/\&#211;/Ó/g;
s/\&Oacute;/Ó/g;
s/\&#212;/Ô/g;
s/\&Ocirc;/Ô/g;
s/\&#213;/Õ/g;
s/\&Otilde;/Õ/g;
s/\&#214;/Ö/g;
s/\&Ouml;/Ö/g;
s/\&#215;/×/g;
s/\&times;/×/g;
s/\&#216;/Ø/g;
s/\&Oslash;/Ø/g;
s/\&#217;/Ù/g;
s/\&Ugrave;/Ù/g;
s/\&#218;/Ú/g;
s/\&Uacute;/Ú/g;
s/\&#219;/Û/g;
s/\&Ucirc;/Û/g;
s/\&#220;/Ü/g;
s/\&Uuml;/Ü/g;
s/\&#221;/Ý/g;
s/\&Yacute;/Ý/g;
s/\&#222;/Þ/g;
s/\&THORN;/Þ/g;
s/\&#223;/ß/g;
s/\&szlig;/ß/g;
s/\&#224;/à/g;
s/\&agrave;/à/g;
s/\&#225;/á/g;
s/\&aacute;/á/g;
s/\&#226;/â/g;
s/\&acirc;/â/g;
s/\&#227;/ã/g;
s/\&atilde;/ã/g;
s/\&#228;/ä/g;
s/\&auml;/ä/g;
s/\&#229;/å/g;
s/\&aring;/å/g;
s/\&#230;/æ/g;
s/\&aelig;/æ/g;
s/\&#231;/ç/g;
s/\&ccedil;/ç/g;
s/\&#232;/è/g;
s/\&egrave;/è/g;
s/\&#233;/é/g;
s/\&eacute;/é/g;
s/\&#234;/ê/g;
s/\&ecirc;/ê/g;
s/\&#235;/ë/g;
s/\&euml;/ë/g;
s/\&#236;/ì/g;
s/\&igrave;/ì/g;
s/\&#237;/í/g;
s/\&iacute;/í/g;
s/\&#238;/î/g;
s/\&icirc;/î/g;
s/\&#239;/ï/g;
s/\&iuml;/ï/g;
s/\&#240;/ð/g;
s/\&ieth;/ð/g;
s/\&#241;/ñ/g;
s/\&ntilde;/ñ/g;
s/\&#242;/ò/g;
s/\&ograve;/ò/g;
s/\&#243;/ó/g;
s/\&oacute;/ó/g;
s/\&#244;/ô/g;
s/\&ocirc;/ô/g;
s/\&#245;/õ/g;
s/\&otilde;/õ/g;
s/\&#246;/ö/g;
s/\&ouml;/ö/g;
s/\&#247;/÷/g;
s/\&divide;/÷/g;
s/\&#248;/ø/g;
s/\&oslash;/ø/g;
s/\&#249;/ù/g;
s/\&ugrave;/ù/g;
s/\&#250;/ú/g;
s/\&uacute;/ú/g;
s/\&#251;/û/g;
s/\&ucirc;/û/g;
s/\&#252;/ü/g;
s/\&uuml;/ü/g;
s/\&#253;/ý/g;
s/\&yacute;/ý/g;
s/\&#254;/þ/g;
s/\&thorn;/þ/g;
s/\&#255;/ÿ/g;
s/\&yuml;/ÿ/g;
s/\&#34;/"/g;
s/\&quot;/"/g;

# Do the ampersand last, so it won't affect the other substitutions
s/\&#38;/\&/g;
s/\&amp;/\&/g;

	return $_;
}

File: lib/htuml2txt.c
/*  	Program Written to Untag a HTML file : Maqsood Mohammed (25 Jan 2000) */
/*      This is a .c version of htuml2txt.pl file to be incorporated in glimpse */

#include <stdio.h>
#include <string.h>
#include <ctype.h>

char lineToSearch[100];
void fixSpecial(void);
static char* string[]= {"b","/b","i","/i","em","/em","font","/font","strong","/strong","big","/bi
g","sup","/sup","sub","/sub","u","/u","strike","/strike","style","/style"};
static char* STRING[]= {"B","/B","I","/I","EM","/EM","FONT","/FONT","STRONG","/STRONG","BIG","/BI
G","SUP","/SUP","SUB","/SUB","U","/U","STRIKE","/STRIKE","STYLE","/STYLE"};
int check(char *arr);

main()
{
        int c,k,tagLen,tagread;
        char *getLine,*tempstr,*isitspecial;
        FILE *tempfp;
        getLine=(char*)malloc(100);
        tempstr=(char*)malloc(100);
        isitspecial=(char*)malloc(100);
        tempfp=fopen("temp.html","w+");
           /*create a temporary file to store results of untaging input file */

        while(1)
        {
            k=0;tagread=0;tagLen=0;
            c=getc(stdin);
            if (c==EOF) break;
            if (c != '<')
                      putc(c,tempfp);
            else
                while (c != '>') {
                   c=getc(stdin);
                   if (!isspace(c) && (c!='>'))
                   {
                   tagLen++;  if (!tagread) tempstr[k++]=c;
                    }
                   else if (isspace(c))/* we are concerned only with tag entity ,not the whole tag*/
                    {
                      tagLen++; tempstr[k]='\0';
                      tagread=1;strcpy(isitspecial,tempstr);
                    }
                   else if (c == '>') /* end of tag reached*/
                    {
                      tempstr[k]='\0';tagread=1;
                     strcpy(isitspecial,tempstr);
                    }
                   if (c==EOF) break;
                }
             if (!check(isitspecial)&&(tagLen!=0)) /* check if the tag is a special tag,if not put a space */
               putc(' ',tempfp);
        }

        fseek(tempfp,0,SEEK_SET); /* go to start of temporary file */
        while ((fgets(getLine,80,tempfp))!=NULL)
        {
              strcpy(lineToSearch,getLine);
              fixSpecial(); /* fix those special characters by reading from temporary file,which contains the untagged ones */
              strcpy(getLine,lineToSearch);
              fputs(getLine,stdout); /* write the modified line into output file */
        }
        fclose(tempfp);
        remove("temp.html"); /* remove the temporary file */
}

int check(char *array)
{
       int i,flag=0;
       for (i=0;i<22;i++) /* comparision if array is a special tag */
       if(strcmp(array,string[i])==0)
            flag=1;
       toupper(array);
       for (i=0;i<22;i++)
       if(strcmp(array,STRING[i])==0)/* upper case comparision */
            flag=1;
       if (flag) return 1; /* if special tag,return 1 else return 0*/
       else return 0;
}

SearchReplace(char inputString[100],char searchPattern[50],char replacePattern[50])
{

     int i,j,k,patternStartsAt,t,ii,patternFound=0,remLen;
     char tempinputString[100];
     for (i=0;i<strlen(inputString);i++)    /* SEARCH */
     {
           k=0;ii=i;
           for (j=0;j<strlen(searchPattern);ii++,j++)
            if (inputString[ii]==searchPattern[j])
                    k++;
           if (k==strlen(searchPattern)) {
                patternFound=1;
                patternStartsAt=i;
           }
     }
     strcpy(tempinputString,inputString);
     remLen=strlen(tempinputString)-patternStartsAt-strlen(searchPattern);
     if (patternFound)                        /* REPLACE */
     {
          for (i=0;i<strlen(replacePattern);i++)
                  tempinputString[patternStartsAt+i]=replacePattern[i];
          for (k=0;k<remLen;k++)
                  tempinputString[patternStartsAt+i+k]=inputString[patternStartsAt+strlen(searchPattern)+k];
          tempinputString[patternStartsAt+i+k]='\0';
          strcpy(lineToSearch,tempinputString); /* lineToSearch is modified to contain the 'new replaced' line */
     }
}




void fixSpecial(){
       char a,replace[50];


a=' ';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#160;",replace);
       SearchReplace(lineToSearch,"&nbsp;",replace);

a='¡';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#161;",replace);
       SearchReplace(lineToSearch,"&iexcl;",replace);

a='¢';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#162;",replace);
       SearchReplace(lineToSearch,"&cent;",replace);

a='£';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#163;",replace);
       SearchReplace(lineToSearch,"&pound;",replace);

a='¤';replace[0]=a;replace[1]='\0';
         SearchReplace(lineToSearch,"&#164;",replace);
       SearchReplace(lineToSearch,"&curren;",replace);

a='¥';replace[0]=a;replace[1]='\0';
SearchReplace(lineToSearch,"&#165;",replace);
       SearchReplace(lineToSearch,"&yen;",replace);
     
a='¦';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#166;",replace);
       SearchReplace(lineToSearch,"&brvbar;",replace);

a='§';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#167;",replace);
       SearchReplace(lineToSearch,"&sect;",replace);

a='¨';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#168;",replace);
       SearchReplace(lineToSearch,"&uml;",replace);

a='©';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#169;",replace);
       SearchReplace(lineToSearch,"&copy;",replace);

a='ª';replace[0]=a;replace[1]='\0';
SearchReplace(lineToSearch,"&#170;",replace);
       SearchReplace(lineToSearch,"&ordf;",replace);


a='«';replace[0]=a;replace[1]='\0';
     SearchReplace(lineToSearch,"&#171;",replace);
       SearchReplace(lineToSearch,"&laquo;",replace);
  
a='¬';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#172;",replace);
       SearchReplace(lineToSearch,"&not;",replace);

a='\\';replace[0]=a;replace[1]='\0';
    SearchReplace(lineToSearch,"&#173;",replace);
       SearchReplace(lineToSearch,"&shy;",replace);
   
a='®';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#174;",replace);
       SearchReplace(lineToSearch,"&reg;",replace);
     
a='¯';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#175;",replace);
       SearchReplace(lineToSearch,"&macr;",replace);
     

a='°';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#176;",replace);
       SearchReplace(lineToSearch,"&deg;",replace);
     

a='±';replace[0]=a;replace[1]='\0';
     SearchReplace(lineToSearch,"&#177;",replace);
       SearchReplace(lineToSearch,"&plusmn;",replace);
     

a='²';replace[0]=a;replace[1]='\0';
SearchReplace(lineToSearch,"&#178;",replace);
       SearchReplace(lineToSearch,"&sup2;",replace);
     


a='³';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#179;",replace);
       SearchReplace(lineToSearch,"&sup3;",replace);
    

a='´';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#180;",replace);
       SearchReplace(lineToSearch,"&acute;",replace);
      

a='µ';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#181;",replace);
       SearchReplace(lineToSearch,"&micro;",replace);

a='¶';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#182;",replace);
       SearchReplace(lineToSearch,"&para;",replace);
     

a='·';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#183;",replace);
       SearchReplace(lineToSearch,"&middot;",replace);
     

a='¸';replace[0]=a;replace[1]='\0';
    SearchReplace(lineToSearch,"&#184;",replace);
       SearchReplace(lineToSearch,"&cedil;",replace);
   
a='¹';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#185;",replace);
       SearchReplace(lineToSearch,"&sup1;",replace);
     

a='º';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#186;",replace);
       SearchReplace(lineToSearch,"&ordm;",replace);
     
a='»';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#187;",replace);
       SearchReplace(lineToSearch,"&raquo;",replace);
     

a='¼';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#188;",replace);
       SearchReplace(lineToSearch,"&frac14;",replace);
    

a='½';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#189;",replace);
       SearchReplace(lineToSearch,"&frac12;",replace);
     

a='¾';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#190;",replace);
       SearchReplace(lineToSearch,"&frac34;",replace);
   

a='¿';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#191;",replace);
       SearchReplace(lineToSearch,"&iquest;",replace);
     

a='À';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#192;",replace);
       SearchReplace(lineToSearch,"&Agrave;",replace);
     
a='Á';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#193;",replace);
       SearchReplace(lineToSearch,"&Aacute;",replace);
    

a='Â';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#194;",replace);
       SearchReplace(lineToSearch,"&circ;",replace);
      

a='Ã';replace[0]=a;replace[1]='\0';
    SearchReplace(lineToSearch,"&#195;",replace);
       SearchReplace(lineToSearch,"&Atilde;",replace);
   

a='Ä';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#196;",replace);
       SearchReplace(lineToSearch,"&Auml;",replace);
    
a='Å';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#197;",replace);
       SearchReplace(lineToSearch,"&ring;",replace);
      

a='Æ';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#198;",replace);
       SearchReplace(lineToSearch,"&AElig;",replace);
    

a='Ç';replace[0]=a;replace[1]='\0';
    SearchReplace(lineToSearch,"&#199;",replace);
       SearchReplace(lineToSearch,"&Ccedil;",replace);
    

a='È';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#200;",replace);
       SearchReplace(lineToSearch,"&Egrave;",replace);
      

a='É';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#201;",replace);
       SearchReplace(lineToSearch,"&Eacute;",replace);
     

a='Ê';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#202;",replace);
       SearchReplace(lineToSearch,"&Ecirc;",replace);
     

a='Ë';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#203;",replace);
       SearchReplace(lineToSearch,"&Euml;",replace);
     

a='Ì';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#204;",replace);
       SearchReplace(lineToSearch,"&Igrave;",replace);
     

a='Í';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#205;",replace);
       SearchReplace(lineToSearch,"&Iacute;",replace);
    

a='Î';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#206;",replace);
       SearchReplace(lineToSearch,"&Icirc;",replace);
     

a='Ï';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#207;",replace);
       SearchReplace(lineToSearch,"&Iuml;",replace);
     

a='Ð';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#208;",replace);
       SearchReplace(lineToSearch,"&ETH;",replace);
      

a='Ñ';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#209;",replace);
       SearchReplace(lineToSearch,"&Ntilde;",replace);
    

a='Ò';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#210;",replace);
       SearchReplace(lineToSearch,"&Ograve;",replace);
     

a='Ó';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#211;",replace);
       SearchReplace(lineToSearch,"&Oacute;",replace);
     

a='Ô';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#212;",replace);
       SearchReplace(lineToSearch,"&Ocirc;",replace);
     

a='Õ';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#213;",replace);
       SearchReplace(lineToSearch,"&Otilde;",replace);
     

a='Ö';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#214;",replace);
       SearchReplace(lineToSearch,"&Ouml;",replace);

a='×';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#215;",replace);
       SearchReplace(lineToSearch,"&times;",replace);
     

a='Ø';replace[0]=a;replace[1]='\0';
    SearchReplace(lineToSearch,"&#216;",replace);
       SearchReplace(lineToSearch,"&Oslash;",replace);
    

a='Ù';replace[0]=a;replace[1]='\0';
SearchReplace(lineToSearch,"&#217;",replace);
       SearchReplace(lineToSearch,"&Ugrave;",replace);
      

a='Ú';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#218;",replace);
       SearchReplace(lineToSearch,"&Uacute;",replace);
    

a='Û';replace[0]=a;replace[1]='\0';
    SearchReplace(lineToSearch,"&#219;",replace);
       SearchReplace(lineToSearch,"&Ucirc;",replace);
   

a='Ü';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#220;",replace);
       SearchReplace(lineToSearch,"&Uuml;",replace);
     

a='Ý';replace[0]=a;replace[1]='\0';
     SearchReplace(lineToSearch,"&#221;",replace);
       SearchReplace(lineToSearch,"&Yacute;",replace);
   

a='Þ';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#222;",replace);
       SearchReplace(lineToSearch,"&THORN;",replace);
      

a='ß';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#223;",replace);
       SearchReplace(lineToSearch,"&szlig;",replace);
      

a='à';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#224;",replace);
       SearchReplace(lineToSearch,"&agrave;",replace);
     

a='á';replace[0]=a;replace[1]='\0';
      SearchReplace(lineToSearch,"&#225;",replace);
       SearchReplace(lineToSearch,"&aacute;",replace);
 

a='â';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#226;",replace);
       SearchReplace(lineToSearch,"&acirc;",replace);
     

a='ã';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#227;",replace);
       SearchReplace(lineToSearch,"&atilde;",replace);
    

a='ä';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#228;",replace);
       SearchReplace(lineToSearch,"&auml;",replace);
     

a='å';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#229;",replace);
       SearchReplace(lineToSearch,"&aring;",replace);
     

a='æ';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#230;",replace);
       SearchReplace(lineToSearch,"&aelig;",replace);
    

a='ç';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#231;",replace);
       SearchReplace(lineToSearch,"&ccedil;",replace);
    

a='è';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#232;",replace);
       SearchReplace(lineToSearch,"&egrave;",replace);
    

a='é';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#233;",replace);
       SearchReplace(lineToSearch,"&eacute;",replace);
    

a='ê';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#234;",replace);
       SearchReplace(lineToSearch,"&ecirc;",replace);
     

a='ë';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#235;",replace);
       SearchReplace(lineToSearch,"&euml;",replace);
     

a='ì';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#236;",replace);
       SearchReplace(lineToSearch,"&igrave;",replace);
     

a='í';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#237;",replace);
       SearchReplace(lineToSearch,"&iacute;",replace);
     

a='î';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#238;",replace);
       SearchReplace(lineToSearch,"&icirc;",replace);
    

a='ï';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#239;",replace);
       SearchReplace(lineToSearch,"&iuml;",replace);
    

a='ð';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#240;",replace);
       SearchReplace(lineToSearch,"&ieth;",replace);
    

a='ñ';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#241;",replace);
       SearchReplace(lineToSearch,"&ntilde;",replace);
     

a='ò';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#242;",replace);
       SearchReplace(lineToSearch,"&ograve;",replace);
    

a='ó';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#243;",replace);
       SearchReplace(lineToSearch,"&oacute;",replace);
      

a='ô';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#244;",replace);
       SearchReplace(lineToSearch,"&ocirc;",replace);
   

a='õ';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#245;",replace);
       SearchReplace(lineToSearch,"&otilde;",replace);
     

a='ö';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#246;",replace);
       SearchReplace(lineToSearch,"&ouml;",replace);
    

a='÷';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#247;",replace);
       SearchReplace(lineToSearch,"&divide;",replace);
      

a='ø';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#248;",replace);
       SearchReplace(lineToSearch,"&oslash;",replace);
      

a='ù';replace[0]=a;replace[1]='\0';
    SearchReplace(lineToSearch,"&#249;",replace);
       SearchReplace(lineToSearch,"&ugrave;",replace);
   

a='ú';replace[0]=a;replace[1]='\0';
 SearchReplace(lineToSearch,"&#250;",replace);
       SearchReplace(lineToSearch,"&uacute;",replace);
    

a='û';replace[0]=a;replace[1]='\0';
    SearchReplace(lineToSearch,"&#251;",replace);
       SearchReplace(lineToSearch,"&ucirc;",replace);
   

a='ü';replace[0]=a;replace[1]='\0';
  SearchReplace(lineToSearch,"&#252;",replace);
       SearchReplace(lineToSearch,"&uuml;",replace);
     

a='ý';replace[0]=a;replace[1]='\0';
   SearchReplace(lineToSearch,"&#253;",replace);
       SearchReplace(lineToSearch,"&yacute;",replace);

a='þ';replace[0]=a;replace[1]='\0';
       SearchReplace(lineToSearch,"&#254;",replace);
       SearchReplace(lineToSearch,"&thorn;",replace);

a='ÿ';replace[0]=a;replace[1]='\0';
       SearchReplace(lineToSearch,"&#255;",replace);
       SearchReplace(lineToSearch,"&yuaml;",replace);

a='"';replace[0]=a;replace[1]='\0';
       SearchReplace(lineToSearch,"&#34;",replace);
       SearchReplace(lineToSearch,"&quot;",replace);

a='&';replace[0]=a;replace[1]='\0';
       SearchReplace(lineToSearch,"&#38;",replace);
       SearchReplace(lineToSearch,"&amp;",replace); 

}
